ethcore/sync/src/chain/supplier.rs

```diff
@@ -387,21 +420,37 @@ impl SyncSupplier {
 				Ok(bytes) => {
 					let mut rlp = RlpStream::new_list(1);
 					rlp.append(&bytes);
-					Ok(Some((PrivateStatePacket.id(), rlp)))
+					Ok(Some((PrivateStatePacket, rlp)))
 				}
 			}
 		})
 	}
 
-	fn return_rlp<FRlp, FError>(io: &mut dyn SyncIo, rlp: &Rlp, peer: PeerId, rlp_func: FRlp, error_func: FError) -> Result<(), PacketDecodeError>
+	fn return_rlp<FRlp, FError>(io: &mut dyn SyncIo, rlp: &Rlp, peer: PeerId, rlp_func: FRlp, error_func: FError) -> Result<(), PacketProcessError>
+		where FRlp : Fn(&dyn SyncIo, &Rlp, PeerId) -> RlpResponseResult,
+			FError : FnOnce(network::Error) -> String
+	{
+		let response = rlp_func(io, rlp, peer);
+		match response {
+			Err(e) => Err(e),
+			Ok(Some((packet_id, rlp_stream))) => {
+				io.respond(packet_id.id(), rlp_stream.out()).unwrap_or_else(
+					|e| debug!(target: "sync", "{:?}", error_func(e)));
+				Ok(())
+			}
+			_ => Ok(())
+		}
```

Comment:
```suggestion
		if let Some(packet_id, rlp_stream) = response? {
			io.respond(packet_id.id(), rlp_stream.out()).unwrap_or_else(
				|e| debug!(target: "sync", "{:?}", error_func(e)));
		}
		Ok(())
```

nit: That's a tiny bit shorter.

Reply:
No )) It's shorter because you omitted Err processing.

Reply:
the error is returned with `?` after `response`, but TBH, it makes the code harder to read, so I would be in favor of keeping it as is

Reply:
Aaa, sorry I missed this ?